server/controllers/users.js

```diff
@@ -0,0 +1,198 @@
+import Helpers from './helpers';
+
+/**
+ * Secret token for jsonwebtoken
+ */
+const secret = process.env.SECRET || 'my secret key';
+
+const usersController = {
+  /**
+   * Login a user
+   * @param {Object} req - Request object
+   * @param {Object} res - Response object
+   * @returns {Object} Response object
+   */
+  login(req, res) {
+    db.User.findOne({ where: { email: req.body.email } })
+      .then((user) => {
+        if (user) {
+          if (user.validatePassword(req.body.password)) {
+            const token = jwt.sign({ userId: user.id, roleId: user.roleId}, 
+            secret, { expiresIn: '1 day' });
+            return res.status(200).send({token, user});
+          }
+        }
+        res.status(401)
+            .send({ message: 'Invalid Login Details!' });
+      });
+  },
+
+ /**
+   * Create a user
+   * @param {Object} req - Request object
+   * @param {Object} res - Response object
+   * @returns {Object} Response object
+   */
+  create(req, res) {
+    db.User.findOne({ where: { email: req.body.email } })
+      .then((existingUser) => {
+        if (existingUser) {
+          return res.status(400).send({ 
+            message: `Oops! A user already exists with this email: ${req.body.email}` 
+          });
+        }
+
+        if (!req.body.roleId) {
+          req.body.roleId = 2;
+        }
+
+        db.User
+          .create({
+            firstname: req.body.firstname,
+            lastname: req.body.lastname,
+            email: req.body.email,
+            password: req.body.password,
+            roleId: req.body.roleId,
+          })
+          .then((user) => {
+            const token = jwt.sign({ userId: user.id, roleId: user.roleId}, 
+            secret, { expiresIn: '1 day' });
+            res.status(201).send({token, user});
+          })
+          .catch(error => res.status(400).send(error));
+        });
+  },
+
+   /**
+   * List all users
+   * @param {Object} req - Request object
+   * @param {Object} res - Response object
+   * @returns {Object} Response object
+   */
+  list(req, res) {
+    let query = {};
+    query.limit = Number(req.query.limit) !== 'NaN'? req.query.limit : 10;
+    query.offset = Number(req.query.limit) !== 'NaN'? req.query.offset : 0;
+    db.User
+      .findAll(query)
+      .then(users => res.status(200).send(users));
+  },
+
+   /**
+   * Retrive a user's details
+   * @param {Object} req - Request object
+   * @param {Object} res - Response object
+   * @returns {Object} Response object
+   */
+  retrieveUser(req, res) {
+    db.User
+      .findById(req.params.id)
+      .then((user) => {
+        if (!user) {
+          return res.status(404).send({
+            message: 'User Not Found',
+          });
+        }
+        if (Helpers.isAdmin(req) || Helpers.isCurrentUser(req, user.id)) {
+          res.status(200).send(user);
+        } else {
+          res.status(403)
+            .send({ message: 'You can only retrieve your information!' });
```

Comment:
Dito more generic message. `You are not authorized to view this page` or something.